iocore/net/I_NetVConnection.h

```diff
@@ -643,6 +646,103 @@ class NetVConnection : public AnnotatedVConnection
   NetVConnection(const NetVConnection &) = delete;
   NetVConnection &operator=(const NetVConnection &) = delete;
 
+  enum class ProxyProtocolVersion {
+    UNDEFINED,
+    V1,
+    V2,
+  };
+
+  enum ProxyProtocolData_t {
+    PROXY_PROTO_UNDEFINED,
+    PROXY_PROTO_SRC,
+    PROXY_PROTO_DST,
+  };
+
+  int
+  set_proxy_protocol_addr(ProxyProtocolData_t src_or_dst, ts::string_view &ip_port_str)
+  {
+    int ret = -1;
+    if (src_or_dst == PROXY_PROTO_SRC) {
+      ret = ats_ip_pton(ip_port_str, &pp_info.src_addr);
+    } else {
+      ret = ats_ip_pton(ip_port_str, &pp_info.dst_addr);
+    }
+    return ret;
+  }
+
+  void
+  set_proxy_protocol_src_addr(ts::string_view src)
+  {
+    set_proxy_protocol_addr(PROXY_PROTO_SRC, src);
+  }
+
+  void
+  set_proxy_protocol_dst_addr(ts::string_view src)
+  {
+    set_proxy_protocol_addr(PROXY_PROTO_DST, src);
+  }
+
+  void
+  set_proxy_protocol_src_port(uint16_t val)
+  {
+    pp_info.src_port = val;
+  }
+
+  void
+  set_proxy_protocol_dst_port(uint16_t val)
+  {
+    pp_info.dst_port = val;
+  }
+
+  void
+  set_proxy_protocol_version(const ProxyProtocolVersion ver)
+  {
+    pp_info.proxy_protocol_version = ver;
+  }
+
+  ProxyProtocolVersion
+  get_proxy_protocol_version()
+  {
+    return pp_info.proxy_protocol_version;
+  }
+
+  sockaddr const *get_proxy_protocol_addr(ProxyProtocolData_t);
+
+  sockaddr const *
+  get_proxy_protocol_src_addr()
+  {
+    return get_proxy_protocol_addr(PROXY_PROTO_SRC);
+  }
+
+  uint16_t
+  get_proxy_protocol_src_port()
+  {
+    return ats_ip_port_host_order(this->get_proxy_protocol_addr(PROXY_PROTO_SRC));
+  }
+
+  sockaddr const *
+  get_proxy_protocol_dst_addr()
+  {
+    return get_proxy_protocol_addr(PROXY_PROTO_DST);
+  }
+
+  uint16_t
+  get_proxy_protocol_dst_port()
+  {
+    return ats_ip_port_host_order(this->get_proxy_protocol_addr(PROXY_PROTO_DST));
+  };
+
+  typedef struct _ProxyProtocol {
+    ProxyProtocolVersion proxy_protocol_version = ProxyProtocolVersion::UNDEFINED;
+    uint16_t ip_family;
```

Comment:
If @zwoop is going to nitpick, I'll note these should be `in_port_t`.

Reply:
A more serious question - you have the port values broken out here, but later you have things like
```
(ats_is_ip(&pp_info.src_addr) && ats_ip_port_cast(&pp_info.src_addr)
```
which treats the port as being in the `IpEndpoint`. Why have `src_port` and `dst_port` at all? Are these values different from the port in the `IpEndpoint`?